[WPE] WPE Platform: Opaque region support

carlosgcampos opened the pull request:
#### 6b96e27c62baf1a2d8040ec73bb71c4c4e847030
<pre>
[WPE] WPE Platform: Opaque region support
<a href="https://bugs.webkit.org/show_bug.cgi?id=267705">https://bugs.webkit.org/show_bug.cgi?id=267705</a>

Reviewed by Alejandro G. Castro.

Add API to set the opaque region on a WPEView and use it from WebKit to
set the web view region as opaque when the background color is opaque.

* Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
(webkit_web_view_set_background_color):
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPERectangle.cpp: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
(wpe_rectangle_copy):
(wpe_rectangle_free):
* Source/WebKit/WPEPlatform/wpe/WPERectangle.h: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_set_opaque_rectangles):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandConstructed):
(wpeViewWaylandSetOpaqueRectangles):
(wpe_view_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:

Canonical link: <a href="https://commits.webkit.org/275684@main">https://commits.webkit.org/275684@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/57777314845ac85694237bf4e90bbfbe95784357

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44927 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45134 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38650 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24791 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18911 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/35207 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43104 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18470 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/36608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/16156 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/16104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46615 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38708 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/37990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41899 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17348 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/14286 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/40512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18966 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19030 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5739 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18611 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

psaavedra commented on Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
```diff
@@ -225,7 +229,17 @@ void webkit_web_view_set_background_color(WebKitWebView* webView, WebKitColor* b
+    if (auto* view = static_cast<WebKit::PageClientImpl&>(page.pageClient()).wpeView()) {
+        if (color.isOpaque()) {
+            WPERectangle rect { 0, 0, wpe_view_get_width(view), wpe_view_get_height(view) };
+            wpe_view_set_opaque_rectangles(view, &rect, 1);
```
If I understood well, in general the normal case is setting a window size opaque region due to ["The web view default background color is opaque white, so set the whole view region as opaque initally"](https://github.com/WebKit/WebKit/pull/25344/files#diff-f0d3765a74324105f31da7e43f65412d783bc30f5440777fa09ed47a12d471c0R420).So we only deal with transparent regions when there is a transparent color in background.

carlosgcampos replied:
Right

psaavedra commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -408,8 +414,12 @@ static void wpeViewWaylandConstructed(GObject* object)
+        wpe_view_set_scale(view, scale);
     }
+
+    // The web view default background color is opaque white, so set the whole view region as opaque initally.
```
```suggestion
    // The web view default background color is opaque white, so set the whole view region as opaque initially.
```

psaavedra replied:
Everything looks good to me.

psaavedra commented on Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
```diff
@@ -835,3 +835,23 @@ GList* wpe_view_get_preferred_dma_buf_formats(WPEView* view)
+/**
+ * wpe_view_set_opaque_rectangles:
+ * @view: a #WPEView
+ * @rects: (nullable) (array length=n_rects): opaque rectangles in view-local coordinates
```
Why do we need a list of rects if we are only appending on rectangle in the code?

carlosgcampos replied:
Because the app might have more information when using a transparent background and can provide a different region.

psaavedra replied:
OK, then this is looking forward. For the future.

alexgcastro commented on Source/WebKit/WPEPlatform/wpe/WPERectangle.cpp:
```diff
@@ -0,0 +1,55 @@
+ * @width: The width of the rectangle.
+ * @height: The height of the rectangle.
+ *
+ * Boxed type representing a rectangle with integer coordiantes.
```
typo: coordiantes -> coordinates

alexgcastro reviewed (approved):
LGTM!

carlosgcampos committed:
[WPE] WPE Platform: Opaque region support
https://bugs.webkit.org/show_bug.cgi?id=267705

Reviewed by Alejandro G. Castro.

Add API to set the opaque region on a WPEView and use it from WebKit to
set the web view region as opaque when the background color is opaque.

* Source/WebKit/UIProcess/API/wpe/WebKitWebViewWPE.cpp:
(webkit_web_view_set_background_color):
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPERectangle.cpp: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
(wpe_rectangle_copy):
(wpe_rectangle_free):
* Source/WebKit/WPEPlatform/wpe/WPERectangle.h: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
* Source/WebKit/WPEPlatform/wpe/WPEView.cpp:
(wpe_view_set_opaque_rectangles):
* Source/WebKit/WPEPlatform/wpe/WPEView.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandConstructed):
(wpeViewWaylandSetOpaqueRectangles):
(wpe_view_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:

Canonical link: https://commits.webkit.org/275684@main
